dali/pipeline/operator/error_reporting.h

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATOR_ERROR_REPORTING_H_
+#define DALI_PIPELINE_OPERATOR_ERROR_REPORTING_H_
+
+#include <exception>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+#include "dali/pipeline/data/types.h"
+#include "dali/pipeline/operator/op_spec.h"
+
+namespace dali {
+
+/**
+ * @brief Simple structure capturing the exception to be propagated to the user with additional
+ * context information (location of the error) and message.
+ */
+struct ErrorInfo {
+  std::exception_ptr exception;
+  std::string context_info;
+  std::string additional_message;
+};
+
+void PropagateError(ErrorInfo error);
+
+
+/**
+ * @brief Base error class for exceptions produced in the DALI pipeline. Subtypes should be used
+ * for specific errors that will be mapped into appropriate Python error types of matching names.
+ *
+ * DALI executor automatically extends those kind of errors when they are thrown from operators
+ * and the origin information is present.
+ */
+class DaliError : public std::exception {
```

Comment:
We hardly ever use `Dali` in our source base (less than 10% of occurrences vs DALI - and that excludes macros).